### PR TITLE
[New Connector] Jira connector - On Cloud and Server

### PR DESCRIPTION
**Pull request by moxarth-elastic**
## PR Content

Jira connector with the support for both on-cloud and on-prem platforms.
- Ping functionality with unit test cases.
- Indexing functionality with configured connection string.
- E2E test cases using Flask.

## Checklists

 - [x] Covered the changes with automated tests
 - [x] Tested the changes locally
 - [x] Add a module or a directory in [connectors/sources](https://github.com/elastic/connectors-python/tree/main/connectors/sources)
 - [x] Implement a class that implements all methods described in connectors.source.BaseDataSource
 - [x] Add a unit test in [connectors/sources/tests](https://github.com/elastic/connectors-python/tree/main/connectors/sources/tests) with +90% coverage
 - [x] Declare your connector in [config.yml](https://github.com/elastic/connectors-python/blob/main/config.yml) in the sources section
 - [x] Declare your dependencies in [requirements](https://github.com/elastic/connectors-python/tree/main/requirements). Make sure you pin these dependencies
 - [x] Make sure you use an async lib for your source. If not possible, make sure you don't block the loop
 - [x] When possible, provide a docker image that runs the backend service, so we can test the connector. If you can't provide a docker image, provide the credentials needed to run against an online service.
 - [x] The test backend needs to return more than 10k documents due to 10k being a default size limit for Elasticsearch pagination. Having more than 10k documents returned from the test backend will help the testing connector more deeply.
 - [x] Added functional test to the nightly buildkite

## New Dependency

- [x] New external service dependency added i.e. Flask to run E2E tests.

### CODE REVIEW

File: connectors/source.py
Comment (artem-shelkovnikov) [inline connectors/source.py:254]: This function should go to `utilities` module.

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:336]: This should not happen in `ping` - it's not used in this function and `get_docs` depends on it. You should initialise it in such a way, that both `ping` and `get_docs` could work if called individually, not sequentially.

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:96]: This default value should be inlined

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:126]: This default value should be inlined

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:136]: This default value should be inlined

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:141]: This default value should be inlined

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:152]: Isn't this function a validation function + assignment? I believe it could be partially moved into `_validate_configuration` (validation part), other part can go to initialiser.

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:192]: More information could help - what's the session about? What is it responsible for? What can go wrong with session and be helpful to see in logs to trace problems?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:224]: This docstring seems to be irrelevant to the function, can you fix it?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:227]: What's `url_name`? Description says it's a "query name", name is "url_name", what name is this actually?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:228]: Why can it be not required?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:296]:
```suggestion
            logger.debug(f"{attachment_name} is not supported by TIKA, skipping")
```

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:285]: Why have this flag?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:301]:
```suggestion
                f"File size {attachment_size} of file {attachment_name} is larger than {FILE_SIZE_LIMIT} bytes. Discarding file content"
```

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:329]: Why do this logic with temporary files? I see that you first download the content into file, then on line 325 you read this file into memory fully, so why have a temporary file in the first place?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:334]: Same as below - if it's used in `get_docs`, it should not get initialised in `ping`

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:427]: Something feels weird in this function - framework is able to handle getting attachments from 3rd-party system, all you need to do in your `get_docs` is call `yield document, partial(self.get_content, document)` and framework will call the partial function you provided to download the document and update the document content with extracted content.

Why did you do it differently here?

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/jira.py:464]: I'm confused with this function, why is it done this way? What is it trying to do?

File: connectors/sources/tests/test_jira.py
Comment (artem-shelkovnikov) [inline connectors/sources/tests/test_jira.py:486]: Ideally you should test multiple cases with different data types returned here, it's your main function that you implement and it has only single test for a very basic scenario, can you make sure you test the method properly?

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:336]: Understood your concern here. But, the connector is designed in such a way that `ping` will always gets executed first and then the `get_docs` method. So, there would not be a chance of the `get_docs` method to be called individually since the framework itself is not allowing it. However, once we have a conclusion on this [comment](https://github.com/elastic/connectors-python/pull/493#issuecomment-1438452169), we'll update the code accordingly.

Comment (artem-shelkovnikov) [reply connectors/sources/jira.py:336]: It's not designed this way, it works this way now because it was the most convenient way to do. It might and will change in future, so we want to build connectors that are resilient to changes.

In short, you should treat `ping` and `get_docs()` as methods that can be called in any sequence OR even on their own, all these 3 cases are valid:

1. `ping(); get_docs()`
2. `ping(); ping()`
3. `get_docs()`

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:227]: Oh, good catch! It's the url_name, not query name. I'll update the description.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:228]: Since, Atlassian APIs does not provide pagination for fetching projects and returns all projects at a time. Also, pagination is not needed for ping since it's using `\myself` endpoint. So, it is only required for fetching the Jira issues.

Comment (artem-shelkovnikov) [reply connectors/sources/jira.py:228]: In this case I advise to separate this method into two - one with pagination, one without. They can both use same private methods to simplify logic.

Flags that control the flow of method that drastically is normally an anti-pattern

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:336]: Oh! got it. However this is related to this [comment](https://github.com/elastic/connectors-python/pull/493#issuecomment-1438452169) so do you want me to change this before getting the conclusion on it, or should wait for it? Till the time, i'll address other comments.

Comment (artem-shelkovnikov) [reply connectors/sources/jira.py:336]: I'll try to address the topic in the comment in the mean time, I'll try to get back to you asap

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:427]: This is a way that was suggested by Tarek to achieve the concurrency at the connector level. You can refer this Tarek's comment - https://github.com/elastic/connectors-python/issues/256#issuecomment-1367643403

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:464]: This is a part of Tarek's suggestion - https://github.com/elastic/connectors-python/issues/256#issuecomment-1367643403. This code is working as a consumer i.e. yielding documents to BYOC. In `get_docs` method, we've made two tasks for fetching projects and issues(including the attachments) and once both tasks are completed, consumer stops.

File: connectors/source.py
Comment (moxarth-elastic) [reply connectors/source.py:254]: Alright, will move it to utils module.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:285]: This flag is used to manage the already indexed attachments. If the attachment has been updated then the content will get extracted (`doit` is True at that time) since the _timestamp has also updated for that attachment else `doit` remains False and content will not get extract.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:329]: This is used to avoid the double-load of files. You can refer this issue - https://github.com/elastic/connectors-python/issues/275

Comment (tarekziade) [reply connectors/sources/jira.py:329]: Discussed with Artem in a meeting. This is the right approach. It can also be generalized at the framework level I added https://github.com/elastic/connectors-python/issues/534

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:96]: We are following a best practice to create a constant for the default values since that's not gonna change and pass it to configurations. So, do you want us to eliminate the constant and directly pass the default value i.e. True for `is_cloud`?

Comment (artem-shelkovnikov) [reply connectors/sources/jira.py:96]: My thought process here:

1. These default-value constants are used in a single place - in `get_default_configuration` method, so they are not re-usable
2. If your IDE does not support tooltips (I use vim, so I'm biased here), then to figure out the value of constant while looking at the configuration, you'll need to scroll to the beginning of file to see the value - that breaks the flow of thoughts/reading slightly.

All in all I just don't see a benefit of moving it to the beginning of the file in a constant instead of using default value. Think of it from this perspective - why labels of configurable fields are not constants too? Why field names are not constants?

To me out of these 3 only field names are worth being constants (as you actually use them in code in multiple places), and all others can just be regular values instead of constants.

File: connectors/sources/jira.py
Comment (artem-shelkovnikov) [reply connectors/sources/jira.py:464]: I see. One small thing -  I think it's nice to test this logic if possible and document what this loop means - it takes quite a while to understand what it does tbh.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:192]: I'll update the docstring and logger to add more info about it.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:96]: Cool, Got it. I'll make the configurations inline.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:336]: Sure, please. Do let me know your opinion on this.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:464]: Sure, I'll document it by adding a comment above this code hoping that's fine for you and add a test case for this logic.

File: connectors/sources/tests/test_jira.py
Comment (moxarth-elastic) [reply connectors/sources/tests/test_jira.py:486]: It would be helpful if you provide some more idea which kind of test you want for this method? Since this test is covering multiple data types.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:228]: I see, thanks for the suggestion. I'll update the code accordingly.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:464]: This logic is already covered in the `test_get_docs` test case.

**Comment by moxarth-elastic** [comment]
@tarekziade did you get a chance to review this PR? If you've any feedbacks, I will address them with the Artem's one.

File: connectors/sources/jira.py
Comment (wangch079) [inline connectors/sources/jira.py:467]: Is this line needed? If you exit the infinite loop above, does it mean both `project_task` and `attachment_task` are done?

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:204]: Yes, this will work for both cloud and on-prem.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:467]: Yes, the consumer stops when it gets `FINISHED` twice from the queue and that is going to happen at the time when both tasks are done.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:145]: Since, the framework by default support 10 concurrent downloads. So, if the product itself allows more concurrent downloads, we can increase it using this config param.

**Comment by moxarth-elastic** [comment]
@wangch079 I've incorporated your comments, can you please give it a moment to review? Also, @tarekziade did you get a chance to review the PR?

File: connectors/sources/jira.py
Comment (vidok) [inline connectors/sources/jira.py:40]: I see you are multiplying this value by 1024*1024 then it is worth mentioning that the value is in megabytes. Another option is to move the multiplication to this line from line 86.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py]: Keeping this to both the places i.e. `ping` and `get_docs`

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:40]: I'll mention that the value is in megabytes through an inline comment.

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/jira.py
Comment (timgrein) [inline connectors/sources/jira.py:195]: I would really like to see the same pattern we've used for the Confluence connector (see: https://github.com/elastic/connectors-python/pull/568/files), i.e. creating a `JiraClient` class, which encapsulates the interaction with the server/cloud instance. Makes the connectors way more consistent.

**Review by timgrein — COMMENTED** [review]
I think we should use the same pattern as in the confluence connector encapsulating and memoizing the client (instance).

File: connectors/sources/jira.py
Comment (wangch079) [reply connectors/sources/jira.py:209]: I just want to confirm, if `self.is_cloud is True`, you are trying to use `service_account_id` and `api_token` to build a basic auth.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:195]: Sure, however the testing cycle has been completed and it would take quite some efforts from the QA end to re-test it with new changes, so I strongly recommend we raise a GitHub issue and cover it as an enhancement PR. Let me know your thoughts regarding the same.

**Comment by tarekziade** [comment]
> I had concerns about how the `get_docs` is implemented.

I'll do a fill review now

> I also think it's overkill to use a queue as well. Having dedicated coroutine for project and issues, not a big deal, but I don't feel it's necessary. cc @tarekziade

`ConcurrentTasks` is just a way to group async tasks together, I don't think it's a problem to do it this way if this source class can boost how fast it's able to browse content from the 3rd party that way and stream them in a queue in a controlled throttled way. Here it's a nice way to browse different kinds of objects in parallel, that all end up being docs (projects and issues).

This technique speeds up the time taken by the connector to grab content.

File: connectors/sources/jira.py
Comment (tarekziade) [inline connectors/sources/jira.py:342]: let's try except this with just a warning if remove failed

**Review by tarekziade — DISMISSED** [review]
just a few nits. LGTM otherwise

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:42]: I've already defined a config parameter for this value - https://github.com/elastic/connectors-python/blob/1c1d6a747221581ba23a713470540a5759cd2527/connectors/sources/jira.py#L148.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:53]: Got it

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:310]: Yes, we can put a warning logger instead.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:435]: Error we're getting here - 

image

**Comment by tarekziade** [comment]
> > `ConcurrentTasks` is just a way to group async tasks together, I don't think it's a problem to do it this way if this source class can boost how fast it's able to browse content from the 3rd party that way and stream them in a queue in a controlled throttled way. Here it's a nice way to browse different kinds of objects in parallel, that all end up being docs (projects and issues).
> 
> FYI @tarekziade , after adding concurrency using `ConcurrentTasks`, we faced `QueueFull` error several times and so the connector would get terminated. So, can we add a `debug` or `warning` logger to show the `QueueFull` message instead of raising it?

The way it works: 
- you will get debug messages on that event (https://github.com/elastic/connectors-python/blob/main/connectors/utils.py#LL229C6-L229C6)
- if the queue stays full for more than 60 seconds, it gets raised (https://github.com/elastic/connectors-python/blob/main/connectors/utils.py#L226)

So if you are raising it means you are filing the queue super fast and the rest of the pipeline can't follow. 

Try this:

```
self.queue = MemQueue(maxsize=0, maxmemsize=10*1024*1024, refresh_timeout=120)
```


o raise it to `120`

File: connectors/sources/jira.py
Comment (tarekziade) [inline connectors/sources/jira.py:283]: if `total` does not exists you get `None` and the next test will fail. 

if you always get `total` and it's an int, change this to:

```
total = response_json["total"]
```

if not, make that code more robust:

```
total = int(response_json.get("total", 0))
```

File: connectors/sources/jira.py
Comment (tarekziade) [inline connectors/sources/jira.py:338]: please add an issue about this so we have the same behavior for `convert_to_b64` on all platform

File: connectors/sources/jira.py
Comment (tarekziade) [inline connectors/sources/jira.py:378]: why do you feed `when` ? it looks like the default value as it's done in `iso_utc`

File: connectors/sources/jira.py
Comment (tarekziade) [inline connectors/sources/jira.py:443]: why `copy` here?

File: connectors/utils.py
Comment (tarekziade) [inline connectors/utils.py:415]: nice!

**Review by tarekziade — COMMENTED** [review]
LGTM, just a few nits to address

**Review by tarekziade — APPROVED** [review]
Approved, pending one more approval since it's a big patch, +my nits to address

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:378]: Since, we're indexing project every time with _timestamp with `iso_utc` and other jira documents are using the timestamp getting from the response. So, if Jira server is hosted in different timezone other than UTC, user will face timestamp mismatch for projects and other objects in term of timezone. So, keep it consitent, we're feeding `when` to the `iso_utc` of the timezone of Jira server.

File: connectors/sources/jira.py
Comment (moxarth-elastic) [reply connectors/sources/jira.py:442]: Here, attachment is a dictionary so any changes in the `partial` should not be reflected in the below code. Hence, we passed the `copy` to avoid changes done by the partial in the passed param.

**Review by timgrein — APPROVED** [review]
LGTM

**Pull request merged**
